### ISSUE TITLE
BUG: `eval_hermite` is incompatible with NumPy 2.0 on Windows

### ISSUE THREAD

**Issue by njzjz** [scipy/scipy#21052]
### Describe your issue.

```
python -c "import scipy.special, numpy as np;print(scipy.special.eval_hermite(np.arange(2), np.zeros((1,2), dtype=int)))"
  
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: ufunc 'eval_hermite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

One can create a GitHub Actions workflow to validate:

```yml
name: Test

on:
  push:

jobs:
  test:
    name: test scipy
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v4
    - uses: actions/setup-python@v5
      with:
        python-version: "3.x"
    - run: pip install scipy
    - run: python -c "import scipy.special, numpy as np;print(scipy.special.eval_hermite(np.arange(2), np.zeros((1,2), dtype=int)))"
```

### Reproducing Code Example

```python
import scipy.special, numpy as np
print(scipy.special.eval_hermite(np.arange(2), np.zeros((1,2), dtype=int)))
```

### Error message

```shell
TypeError: ufunc 'eval_hermite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

### SciPy/NumPy/Python version and system information

```shell
scipy 1.14.0
numpy 2.0.0
python 3.12.4
Microsoft Windows Server 2022
```

**Comment by lucascolley** [comment]
does `scipy.special.eval_hermite(2, np.zeros((1,2), dtype=int))` give the same error?

**Comment by njzjz** [comment]
> does `scipy.special.eval_hermite(2, np.zeros((1,2), dtype=int))` give the same error?

No.

```
Run python -c "import scipy.special, numpy as np;print(scipy.special.eval_hermite(2, np.zeros((1,2), dtype=int)))"
  
[[-2. -2.]]
```

**Comment by andyfaff** [comment]
```
import scipy.special
import numpy as np;
print(scipy.special.eval_hermite(np.arange(2), np.zeros((1,2), dtype=int)))
```

works for me on scipy 1.13.1 and 1.14.0, and numpy 1.26.4 and 2.0.0 on macOS14.5, cp12.

**Comment by tylerjereddy** [comment]
> cdef inline double eval_hermite(long n, double x) noexcept nogil

Probably `long` is a 32-bit integer on win64 here, but NumPy changed its own default integer type on win64 to 64-bit integer in NumPy `2.0.0`. We had something fairly similar to this patched at: https://github.com/scipy/scipy/pull/19724/files#r1433331076.

Let's check with @steppi re: the comment on my patch there:

> I think we could change the ufunc generation infrastructure to explicitly use int64 types instead of long but I don't know offhand how to put everything together to make that work. I think it's fine to just do the conversion in Python for now, and I'll work on updating the ufunc infrastructure before rewriting orthogonal_eval.pxd.

Do we actually have a nice pure Python place to kick the can down the road in this case with `np.dtype("long")`? It wasn't as obvious to me if there's an opportunity for that with this particular code path.

**Comment by steppi** [comment]
> Let's check with @steppi re: the comment on my patch there:

I have a better understanding of how the ufunc generation infrastructure works now, and @izaid and I started to work on replacing it, but we didn't get to this issue with integer types. It can be done in principle though.

Will the problem go away in this case if we just make all of the `long`s here into `int`s? There's no reason for these inputs to be `long`s. Hermite polynomials of degree `2**31` or greater isn't really a sensible thing to compute in finite precision arithmetic.

**Comment by lucascolley** [comment]
x-ref gh-19462, cc @seberg

**Comment by WarrenWeckesser** [comment]
@steppi, are you working on this currently?  I can work on it, but I won't bother if you (or someone else) already have a fix in progress.

**Comment by steppi** [comment]
> @steppi, are you working on this currently? I can work on it, but I won't bother if you (or someone else) already have a fix in progress.

I'm not currently working on this; It looks like I lost track of this one. Thanks, it would be great if you could work on it. In this specific case, I'm pretty sure we can safely change `long` to `int`. 

In the long term, I think the principled solution for the general case would be to change to explicitly using 32 bit and 64 bit ints in the ufuncs; using dtype codes 'i4' and 'i8' in place of `d` and `l` when setting up the ufuncs with the NumPy C API while also changing the types in the scalar kernels. Right now ufuncs are set up in two places, with code generated by `special/_generate_pyx.py`, and directly in `special/_special_ufuncs.cpp`. Changing the latter should be straightforward. I spent 30 minutes looking into doing it for the ufuncs set up in `_generate_pyx.py` (which includes `eval_hermite`) the day I wrote my previous comment in this thread,  but couldn't quite get it working. If you want to try to go for the principled solution, I think you might find that more straightforward than me though.

**Comment by WarrenWeckesser** [comment]
I think `int` will result in the same problem. My understanding is that even on 64 bit Windows, `int` is still 32 bit, but the default NumPy integer type (i.e. what you get from, for example, `np.arange(10)` or `np.array([1, 2, 3])`) is 64 bit.  The ufuncs must be updated to use NumPy's `intp` type.  I'll give it a shot.

**Comment by steppi** [comment]
> I think `int` will result in the same problem. My understanding is that even on 64 bit Windows, `int` is still 32 bit, but the default NumPy integer type (i.e. what you get from, for example, `np.arange(10)` or `np.array([1, 2, 3])`) is 64 bit. The ufuncs must be updated to use NumPy's `intp` type. I'll give it a shot.

I see, that makes sense. I thought because this was only reported for `eval_hermite` that the issue only occured when the input type is `long`. Does the issue also occur for

`nbdtr(n, k, p)` when `n` and `k` are `int`s (signature iid->d)?

I'm traveling now and don't have access to a windows machine to check for myself. 

OK, I see that [intp](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.intp) has character code `n` and should work in a ufunc. That does seem like the right way to go.

**Comment by seberg** [comment]
> OK, I see that [intp](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.intp) has character code "n" and should work in a ufunc.

Right, `n` is not available on NumPy 1.x though, so you'll just have to use `p` instead, which is the same on all relevant platforms.

**Comment by steppi** [comment]
> > OK, I see that [intp](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.intp) has character code "n" and should work in a ufunc.
> 
> Right, `n` is not available on NumPy 1.x though, so you'll just have to use `p` instead, which is the same on all relevant platforms.

Ah yes, of course. Thanks.

**Comment by WarrenWeckesser** [comment]
> I thought because this was only reported for eval_hermite that the issue only occured when the input type is long.

Update: Of the functions tested in the `test_cython_api` in `special/tests/test_cython_special.py`, `eval_hermite` and `eval_hermitenorm` are the only ones where the *only* type defined for a certain input is `long`.  All the other ufuncs with `int` or `long` inputs also handle `double` for the same input, so when they are given a 64 bit integer type on a platform where `long` is 32 bit, the input is cast to `double`.

**Comment by izaid** [comment]
Just a quick comment here, hopefully I've read enough of this thread that I'm saying something sensible.

First, we haven't touched `eval_hermite` yet, so whatever the issue is here predates any of the special overhaul. Second, I did come across the issue of "which NumPy integer type is platform-independent with respect to Python's integer type". I found if I made everything `long long int`, it was all okay. If we don't want to do that, the answer is supporting both 32 bit and 64 bit specifically. But this starts leading to an explosion of type signatures...

**Comment by WarrenWeckesser** [comment]
@seberg, is there a C type that we can be sure is equivalent to `npy_intp`?  Maybe `ptrdiff_t`?

**Comment by WarrenWeckesser** [comment]
> @seberg, is there a C type that we can be sure is equivalent to npy_intp? Maybe ptrdiff_t?

Cython reproduces the type `Py_ssize_t` from the Python API, so that should cover Cython code that wants an integer type that is sure to be compatible with `npy_intp` without having to import numpy headers.

**Comment by seberg** [comment]
[Use `Py_ssize_t](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.intp)?  Otherwise, I guess define your own `ssize_t` as it is non-standard.
Yes, you can use `ptrdiff_t` in practice and that was even correct on NumPy 1.x.  But it would be better not to, because they can differ.
OTOH, if you are doing this to be compatible with the character code `p`, you maybe should use `intptr_t` or so.

Or, maybe you just add the `#ifdef` yourself for whether it is `long` or `long long` (preferring long in NumPy)...

**Comment by WarrenWeckesser** [comment]
It turns out that for the purpose of fixing this issue in https://github.com/scipy/scipy/pull/21401, I was able to use `Py_ssize_t` everywhere I needed an integer that was compatible with `npy_intp`. It is a built-in type in Cython, and in the C++ code, the wrapper header file includes `Python.h`.

**Comment by WarrenWeckesser** [comment]
Fixed in #21401.
